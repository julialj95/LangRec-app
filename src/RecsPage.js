import React, { Component } from "react";
import { Link } from "react-router-dom";
import ResultItem from "./ResultItem";
import TokenService from "./services/token-service";
import config from "./config";
import { LangrecContext } from "./LangrecContext";

class RecsPage extends Component {
  static contextType = LangrecContext;
  constructor() {
    super();
    this.state = {
      language: "",
      type: "",
      level: "",
      cost: "",
      results: [],
      error: "",
      savedResourceIds: [],
      addedToFavorites: false,
    };
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
      error: "",
    });
  };

  displayResults = () => {
    const results = this.context.recommendedResources.map((result) => {
      const isFavorited =
        this.context.savedResourceIds.filter((resource) => {
          return Number(resource) === result.id;
        }).length > 0;
      return (
        <div>
          <ResultItem
            key={result.id}
            id={result.id}
            title={result.title}
            image_link={result.image_link}
            language={result.language}
            level={result.level}
            type={result.type}
            rating={result.rating}
            url={result.url}
            description={result.description}
            cost={result.cost}
          />
          <button
            key={result.title + result.id}
            value={result.id}
            disabled={!this.context.isLoggedIn || isFavorited ? true : false}
            onClick={() => this.saveAResource(result.id)}
          >
            Favorite!
          </button>

          {isFavorited ? (
            <Link to="/saved-resources">View in Saved Resources List</Link>
          ) : null}
        </div>
      );
    });
    return results;
  };

  saveAResource = (resource_id) => {
    fetch(`${config.API_BASE_URL}/resources/recs`, {
      method: "POST",
      body: JSON.stringify({ resource_id }),
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer " + TokenService.getAuthToken(),
      },
    })
      .then((res) => {
        if (!res.ok) {
          res.json().then((e) => Promise.reject(e));
        } else {
          this.context.handleSaveResource(resource_id);

          res.json();
        }
      })
      .catch((error) => console.error({ error }));
  };

  getRecommendations = (e) => {
    e.preventDefault();
    const { language, type, level, cost } = this.state;
    const baseUrl = config.API_BASE_URL;
    const params = `?language=${language}&level=${level}&type=${type}&cost=${cost}`;
    fetch(`${baseUrl}/resources/recs${params}`, {
      headers: {
        "content-type": "application/json",
        // Authorization: "Bearer " + TokenService.getAuthToken(),
      },
    })
      .then((recs) => {
        if (!recs.ok) {
          return recs.json().then((event) => Promise.reject(event));
        }
        return recs.json();
      })
      .then((data) => this.context.handleRecommendedResources(data))
      .catch((error) => {
        this.setState({
          error: "Sorry! No resources match your requirements.",
        });
      });
  };

  render() {
    return (
      <div>
        <h2>Get Reccomendations</h2>
        <h3>
          Enter your information to receive the highest rated resources that fit
          your needs!
        </h3>
        <form onSubmit={(e) => this.getRecommendations(e)}>
          <label htmlFor="language">
            Language
            <select name="language" onChange={(e) => this.handleChange(e)}>
              <option value={null}>Select language...</option>
              <option value="Spanish">Spanish</option>
              <option value="French">French</option>
              <option value="German">German</option>
              <option value="Korean">Korean</option>
              <option value="Mandarin">Mandarin</option>
              <option value="Italian">Italian</option>
              <option value="Cantonese">Cantonese</option>
              <option value="Japanese">Japanese</option>
              <option value="Swahili">Swahili</option>
              <option value="Afrikaans">Afrikaans</option>
              <option value="other">Other</option>
            </select>
          </label>
          <br />
          <label htmlFor="type">
            Resource Type
            <select name="type" onChange={(e) => this.handleChange(e)}>
              <option value={null}>Select Type...</option>
              <option value="Textbook">Textbook</option>
              <option value="Storybook">Storybook</option>
              <option value="Class">Class</option>
              <option value="Website">Website</option>
              <option value="other">Other</option>
            </select>
          </label>
          <br />
          <br />
          <label>
            Level
            <br />
            <input
              type="radio"
              id="beginner"
              name="level"
              value="Beginner"
              onChange={(e) => this.handleChange(e)}
            />
            <label htmlFor="beginner">Beginner</label>
            <br />
            <input
              type="radio"
              id="intermediate"
              name="level"
              value="Intermediate"
              onChange={(e) => this.handleChange(e)}
            />
            <label htmlFor="intermediate">Intermediate</label>
            <br />
            <input
              type="radio"
              id="advanced"
              name="level"
              value="Advanced"
              onChange={(e) => this.handleChange(e)}
            />
            <label htmlFor="advanced">Advanced</label>
          </label>
          <br />
          <br />
          <label>
            Cost
            <br />
            <input
              type="radio"
              name="cost"
              value="Free"
              onChange={(e) => this.handleChange(e)}
            />
            <label htmlFor="cost">Free</label>
            <br />
            <input
              type="radio"
              name="cost"
              value="Paid"
              onChange={(e) => this.handleChange(e)}
            />
            <label htmlFor="cost">Paid</label>
            <br />
          </label>

          <button type="submit">Get Recs!</button>
        </form>

        <div>
          {this.state.error ? <h2>{this.state.error}</h2> : null}
          {this.context.recommendedResources.length === 0 ? null : (
            <div>
              <h1>Results</h1>
              {this.displayResults()}
            </div>
          )}
        </div>
      </div>
    );
  }
}

export default RecsPage;
