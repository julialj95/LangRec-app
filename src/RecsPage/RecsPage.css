@import url("https://fonts.googleapis.com/css2?family=Nunito:wght@300&display=swap");
@import url("https://fonts.googleapis.com/css2?family=Montserrat:wght@300&display=swap");

.recs_block > h1,
h2,
h3,
p {
  font-family: "Montserrat";
  color: white;
}

.results_header {
  font-family: "Montserrat";
  color: white;
}

.recs_form {
  display: flex;
  flex-direction: column;
  margin-left: 10%;
  margin-right: 10%;
}
.form_field {
  background-color: transparent;
  border: 1px solid white;
  border-radius: 5px;
  color: white;
  font-size: 16px;
  font-family: inherit;
  flex-grow: 1;
  height: 40px;
  padding: 0 10px;
  margin: 10px;
  text-align-last: center;
}

.form_field:hover {
  cursor: pointer;
}
.form_field:focus {
  border-color: #1d3557;
  border-width: 2px;
}

button {
  background: linear-gradient(to bottom, #219ebc, #003049 100%);
  border-radius: 5px;
  padding: 5px;
  color: white;
}
button:hover {
  cursor: pointer;
  color: #ff5c8a;
}

.button_row {
  width: 100%;
  text-align: center;
}

option {
  font-family: "Montserrat";
  text-align: center;
  color: darkslategray;
}

/* responsive styling for various sizes */
@media all and (min-width: 600px) {
}
@media all and (min-width: 768px) {
  .recs_form {
    flex-direction: row;
    justify-content: space-around;
  }
}
@media all and (min-width: 992px) {
}

@media all and (min-width: 1200px) {
}
