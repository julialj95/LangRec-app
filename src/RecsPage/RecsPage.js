import React, { Component } from "react";
import ResultItem from "../ResultItem/ResultItem";
import config from "../config";
import { LangrecContext } from "../LangrecContext";
import "./RecsPage.css";

class RecsPage extends Component {
  static contextType = LangrecContext;
  constructor() {
    super();
    this.state = {
      language: "",
      type: "",
      level: "",
      cost: "",
      results: [],
      error: "",
    };
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({
      [name]: value,
      error: null,
    });
  };

  displayResults = () => {
    const results = this.context.recommendedResources.map((result) => {
      const isFavorited =
        this.context.savedResourceIds.filter((resource) => {
          return Number(resource) === result.id;
        }).length > 0;
      return (
        <ResultItem
          key={result.id}
          id={result.id}
          title={result.title}
          image_link={result.image_link}
          language={result.language}
          level={result.level}
          type={result.type}
          rating={result.rating}
          url={result.url}
          description={result.description}
          cost={result.cost}
          favorited={isFavorited}
        />
      );
    });
    return results;
  };

  getRecommendations = (e) => {
    e.preventDefault();
    const { language, type, level, cost } = this.state;
    const baseUrl = config.API_BASE_URL;
    const params = `?language=${language}&level=${level}&type=${type}&cost=${cost}`;

    fetch(`${baseUrl}/resources/recs${params}`, {
      headers: {
        "content-type": "application/json",
      },
    })
      .then((recs) => {
        if (!recs.ok) {
          return recs.json().then((event) => Promise.reject(event));
        }
        return recs.json();
      })
      .then((data) => this.context.handleRecommendedResources(data))
      .catch((error) => {
        this.setState({
          error: "Sorry! No resources match your requirements.",
        });
      });
  };

  render() {
    const length = this.context.recommendedResources.length;
    return (
      <main className="recs_block">
        <h1>GET RECOMMENDATIONS</h1>
        <h2 className="instructions">
          Enter your desired language, type of resource, level and cost
          preference to view the resources that best fit your needs!
        </h2>

        <form onSubmit={(e) => this.getRecommendations(e)}>
          <div className="recs_form">
            <select
              className="form_field"
              name="language"
              onChange={(e) => this.handleChange(e)}
            >
              <option value={null}>Select language...</option>
              <option value="Spanish">Spanish</option>
              <option value="French">French</option>
              <option value="German">German</option>
              <option value="Korean">Korean</option>
              <option value="Mandarin">Mandarin</option>
              <option value="Italian">Italian</option>
              <option value="Cantonese">Cantonese</option>
              <option value="Japanese">Japanese</option>
              <option value="Swahili">Swahili</option>
              <option value="Afrikaans">Afrikaans</option>
              <option value="Portuguese">Portuguese</option>
            </select>

            <br />

            <select
              className="form_field"
              name="type"
              onChange={(e) => this.handleChange(e)}
            >
              <option value={null}>Select Resource Type...</option>
              <option value="Textbook">Textbook</option>
              <option value="Storybook">Storybook</option>
              <option value="Class">Class</option>
              <option value="Website">Website</option>
              <option value="other">Other</option>
            </select>

            <br />
            <select
              className="form_field"
              name="level"
              onChange={(e) => this.handleChange(e)}
            >
              <option value={null}>Select Level...</option>
              <option value="Beginner">Beginner</option>
              <option value="Intermediate">Intermediate</option>
              <option value="Advanced">Advanced</option>
            </select>

            <br />
            <select
              className="form_field"
              name="cost"
              onChange={(e) => this.handleChange(e)}
            >
              <option value={null}>Select Cost...</option>
              <option value="Free">Free</option>
              <option value="Paid">Paid</option>
            </select>
          </div>
          <br />
          <div className="button_row">
            <button className="get_recs_button" type="submit">
              GET RECS
            </button>
          </div>
        </form>

        <section>
          {this.state.error ? (
            <h2>{this.state.error}</h2>
          ) : length > 0 ? (
            <div>
              <h2 className="results_header">RECOMMENDATIONS</h2>
              <div className="results_display">{this.displayResults()}</div>
            </div>
          ) : null}
        </section>
      </main>
    );
  }
}

export default RecsPage;
