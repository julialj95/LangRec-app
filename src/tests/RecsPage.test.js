import React from "react";
import ReactDOM from "react-dom";
import RecsPage from "../RecsPage/RecsPage";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(
    <Router>
      <RecsPage />
    </Router>,
    div
  );

  ReactDOM.unmountComponentAtNode(div);
});

it("renders the UI as expected", () => {
  const tree = renderer
    .create(
      <Router>
        <RecsPage />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
