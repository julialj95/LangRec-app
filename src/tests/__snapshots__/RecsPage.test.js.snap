// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<div
  className="recs_block"
>
  <h2>
    Get Recommendations
  </h2>
  <h3>
    Enter your requirements to view the resources that best fit your needs!
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div
      className="recs_form"
    >
      <select
        className="form_field"
        name="language"
        onChange={[Function]}
      >
        <option
          value={null}
        >
          Select language...
        </option>
        <option
          value="Spanish"
        >
          Spanish
        </option>
        <option
          value="French"
        >
          French
        </option>
        <option
          value="German"
        >
          German
        </option>
        <option
          value="Korean"
        >
          Korean
        </option>
        <option
          value="Mandarin"
        >
          Mandarin
        </option>
        <option
          value="Italian"
        >
          Italian
        </option>
        <option
          value="Cantonese"
        >
          Cantonese
        </option>
        <option
          value="Japanese"
        >
          Japanese
        </option>
        <option
          value="Swahili"
        >
          Swahili
        </option>
        <option
          value="Afrikaans"
        >
          Afrikaans
        </option>
        <option
          value="Portuguese"
        >
          Portuguese
        </option>
      </select>
      <br />
      <select
        className="form_field"
        name="type"
        onChange={[Function]}
      >
        <option
          value={null}
        >
          Select Resource Type...
        </option>
        <option
          value="Textbook"
        >
          Textbook
        </option>
        <option
          value="Storybook"
        >
          Storybook
        </option>
        <option
          value="Class"
        >
          Class
        </option>
        <option
          value="Website"
        >
          Website
        </option>
        <option
          value="other"
        >
          Other
        </option>
      </select>
      <br />
      <select
        className="form_field"
        name="level"
        onChange={[Function]}
      >
        <option
          value={null}
        >
          Select Level...
        </option>
        <option
          value="Beginner"
        >
          Beginner
        </option>
        <option
          value="Intermediate"
        >
          Intermediate
        </option>
        <option
          value="Advanced"
        >
          Advanced
        </option>
      </select>
      <br />
      <select
        className="form_field"
        name="cost"
        onChange={[Function]}
      >
        <option
          value={null}
        >
          Select Cost...
        </option>
        <option
          value="Free"
        >
          Free
        </option>
        <option
          value="Paid"
        >
          Paid
        </option>
      </select>
    </div>
    <br />
    <div
      className="button_row"
    >
      <button
        className="get_recs_button"
        type="submit"
      >
        GET RECS
      </button>
    </div>
  </form>
  <div />
</div>
`;
